src/main/kotlin/com/cultureamp/eventsourcing/RelationalDatabaseEventStore.kt

```diff
@@ -221,16 +226,25 @@ private fun Table.nonUniqueIndex(vararg columns: Column<*>) = index(false, *colu
 object ConcurrencyError : RetriableError
 object LockingError : CommandError
 
-fun Transaction.pgAdvisoryXactLock(): CommandError? {
-    val lockTimeoutMilliseconds = 10_000
-    try {
-        exec("SET LOCAL lock_timeout = '${lockTimeoutMilliseconds}ms';")
-        exec("SELECT pg_advisory_xact_lock(-1)")
-    } catch (e: SQLException) {
-        if (e.message.orEmpty().contains("canceling statement due to lock timeout")) {
-            return LockingError
-        } else {
-            throw e
+fun Transaction.pgAdvisoryXactLock(lockTimeoutMs: Int = 10_000): CommandError? {
+    if (lockTimeoutMs == 0) {
+        // Non-blocking
+        val gotLock = exec("SELECT pg_try_advisory_xact_lock(-1)") { rs ->
+            if (rs.next()) rs.getBoolean(1) else false
+        } ?: false
+
+        if (!gotLock) return LockingError
+    } else {
+        // Blocking with timeout
+        try {
+            exec("SET LOCAL lock_timeout = '${lockTimeoutMs}ms';")
+            exec("SELECT pg_advisory_xact_lock(-1)")
```

Comment:
nitpick: I wonder if we could save a round trip by putting these two together
```suggestion
            exec("SET LOCAL lock_timeout = '${lockTimeoutMs}ms'; SELECT pg_advisory_xact_lock(-1)")
```

Reply:
Oh that's actually sweet - I'll make the change

Reply:
Ok ChatGPT is not sure it's safe, and would need testing:

<img width="806" height="303" alt="image" src="https://github.com/user-attachments/assets/dae5d63a-ad67-49bd-87cc-174a4dde687f" />